Add support for app shell navigation via AMP Shadow DOM

westonruter opened the pull request:
Note: This is somewhat of a prototype for what could be supported in AMP itself as seamless page transitions: https://github.com/ampproject/amphtml/issues/12981 / https://github.com/ampproject/amphtml/issues/14738

----

For a theme to support app shell via the AMP plugin, the bare minimum that needs to be done is:

0. Activate on [PWA plugin](https://wordpress.org/plugins/pwa).
1. Ensure your theme (and plugins) work entirely in the AMP plugin's [Transitional mode](https://amp-wp.org/documentation/how-the-plugin-works/amp-plugin-serving-strategies/). You must use Transitional mode as opposed to Standard mode for now. Make sure that you have “Serve all templates as AMP regardless of what is being queried” enabled.
2. Identify the element that contains the markup that changes as you navigate from page to page and wrap it with two function calls to indicate where the app shell container is:
```php
<?php amp_start_app_shell_content(); ?>
    <?php get_template_part( 'content' ); ?>
<?php amp_end_app_shell_content(); ?>
```
3. Opt-in to AMP theme support for app shell by adding the following to your `functions.php`:

```php
add_theme_support( 'amp', array(
    'paired' => true,
    'app_shell' => true,
) );
```
4. You should define an `offline.php` template in your theme.
5. [Deprecated] ~You should mark for precaching any scripts and stylesheets needed by the app shell. This can be done via:~
```php
wp_style_add_data( 'foo-special-colors', 'precache', true );
wp_script_add_data( 'bar-special-behavior', 'precache', true );
```

For theme which adds support for AMP app shell see https://github.com/xwp/twentyseventeen-westonson. Note that this theme copies some code from the Twenty Seventeen parent theme in order to add make the necessary modifications. For example, the `js/global.js` file modified as per [this diff](https://gist.github.com/westonruter/b9d7952c0879ea3cda9e0081e387846d). The theme has app shell behind a theme mod flag, so make sure you do `wp theme mod set service_worker_navigation amp_app_shell` via WP-CLI. See also how it [adds skeleton content](https://github.com/xwp/twentyseventeen-westonson/blob/03934b7328f9b18dd979ae97458ad38feb6f0e1a/functions.php#L241-L283) to the app shell. See also plugin to [enable stale-while-revalidate caching strategy for navigation requests](https://gist.github.com/westonruter/f848013108672568be6dcde153f9ca37).

This was presented at CDS 2018; see [related portion of talk](https://youtu.be/s1WrBaAyzAI?t=940).

For a demo of the app shell theme in action, see https://dev-showcase-pwa.pantheonsite.io/

See also this short screen cast: https://youtu.be/oAiIbhHdoXM

For some more background on this, see https://github.com/xwp/pwa-wp/issues/12#issuecomment-401843173

# Todo

- [x] Identify the element that contains the content
- [x] Add a query var which is used to obtain the document without the content being populated; add another query var which is used to omit the content for the shell response. Make sure this is done before sanitizers run.
- [x] Preserve admin bar in shadow DOM.
- [x] Add dedicated element which contains the content regardless of the id.
- [x] Ensure all `style` and `link` elements get preserved during content isolation.
- [x] Add Shadow DOM JS in `outer` shell request.
- [x] Let shadow root element be direct child of content element. This avoids styling problems.
- [x] Synchronize `title`.
- [x] Synchronize `body` classes.
- [x] Synchronize nav menu item classes.
- [x] Add JS which is responsible for history management, navigation, fetching, and invoking Shadow DOM API. Intercept the clicks on links in the shell and in the shadow to then load the new shadow doc.
- [x] Bootstrap spa onto frontend, automatically swap out content with shadow DOM AMP doc when navigating to another page? This handles case of smooth transition when navigating after landing on a site for the first time before the service worker is installed. Afterward, the app she would be served instead.
- [x] Fix lack of trailing slash causes redirect which drops the required app shell query param.
- [x] Trigger an event on the document when a page is transitioned.
- [x] Transfer admin bar from shadow to shell document.
- [x] Use AMP Shadow streaming API.
- [x] Update the class name for the menu item the instant that it is clicked.
- [x] ~Flag that the [deprecation of the non-streaming API](https://github.com/ampproject/amphtml/blob/master/spec/amp-shadow-doc.md#fetching-and-attaching-shadow-docs) would break this.~
- [x] Use async external script. Include in precache.
- [x] ~When declaring theme support, include the script handles to print in addition to shadow API script.~
- [x] Make sure that the AMP runtime does not conflict with our custom script which intercept clicks.
- [x] When content is extracted in shell, make sure that shell serves skeleton initially.
- [x] Make sure app shell has revision. Vary cache by navigation menus, theme mods, etc.
- [x] Figure out injecting into service worker for navigation. Disable preload and then override entirely the URL that is served. 
- [x] Make sure app shell is precached, as well as offline and error inners
- [x] Implement two methods of integrating app shell: bootstrapping app shell onto SW-unmanaged page, and another where app shell loads the content from the start.
- [ ] Synchronize meta (`link`, `meta`, JSON-LD, etc)
- [ ] Omit admin bar from shell? Or rather, prevent app shell entirely if user is logged-in?
- [ ] Allow shadow doc request to respond with 400 if amp is disabled and so then cause navigation at top? No, just do not follow redirects from `?amp` to not-`?amp`. If such a redirect happens, navigate the entire page.
- [ ] Stale while revalidate will be able to update. This can be done even on streaming by leaving script in the page which listens for broadcast update.
- [ ] Make sure that the app shell serves the appropriate initial body class based on whether or not the homepage is initially requested.
- [ ] Preserve amp-analytics in shadow DOM; analytics needs to be excluded from app shell.
- [ ] When requesting app shell contents force accepting sanitization
- [ ] How can we handle a case where a user navigate to a URL that has amp disabled.  Add a fury fragment to indicate that the service worker should bypass the use of the app shell
- [ ] Scroll to content of it is out of view on mobile.
- [ ] Prevent navigation when `#` anchor link.
- [ ] Add support for navigating to anchors on other URLs.
- [ ] Download web components API JS to serve locally.

westonruter committed:
Add query var to isolate app shell inner/outer

westonruter committed:
Preserve admin bar for inner shell requests

westonruter commented:
Note that travis is failing due to Gutenberg being merged into core.

westonruter committed:
Fix builds after Gutenberg merged into core

westonruter committed:
Merge branch 'develop' of https://github.com/Automattic/amp-wp into add/spa-amp-shadow-dom

westonruter committed:
Preserve styles when isolating inner content element

westonruter committed:
Inject amp-shadow into outer app shell document

westonruter committed:
Prevent outer app shell doc from being served with amp attribute

westonruter committed:
Serve shadow-v0 instead of v0 as runtime in outer app shells

westonruter committed:
Populate shadow root with content for given URL to test

westonruter committed:
Switch back to using ID for identifying content container

westonruter committed:
Add workaround for :root pesudo selectors in shadow documents

westonruter committed:
Preserve SVG defs when isolating inner content

westonruter committed:
Merge branch 'develop' of https://github.com/Automattic/amp-wp into add/spa-amp-shadow-dom

westonruter committed:
Merge branch 'develop' of https://github.com/Automattic/amp-wp into add/spa-amp-shadow-dom

westonruter committed:
Use template tags to demarcate app shell content

westonruter committed:
Add app shell content placeholder

westonruter committed:
Force outer app shell to be served from non-AMP version

westonruter committed:
Move app shell setup logic into separate method

westonruter committed:
Enqueue amp-shadow and amp-wp-app-shell script in non-inner requests

westonruter committed:
Load AMP into shadow doc when clicking links

westonruter committed:
Preserve purged query vars in canonical redirects

westonruter committed:
Add handling of GET form submissions

westonruter committed:
Upate nav menu class names based on current URL

westonruter committed:
Prevent finishing initialization if not on frontend

westonruter committed:
Let admin bar from shadow doc take over initial app shell admin bar

westonruter committed:
Prevent event default at end of event handler

westonruter committed:
Prevent enqueueing app shell JS in an AMP page

westonruter committed:
Prevent intercepting clicks and form submits in admin bar

westonruter committed:
Prevent loading non-AMP documents into the shadow DOM

westonruter committed:
Improve XHR integration

westonruter committed:
Add popstate support and improve condition for scrolling

westonruter committed:
Merge branch 'develop' of https://github.com/Automattic/amp-wp into add/spa-amp-shadow-dom

westonruter committed:
Triger events for app-shell navigate, load, and ready

westonruter committed:
Do not show fallback source as active theme if no validation errors

westonruter committed:
Temporarily add cache-busting query param for sake of XHR

westonruter committed:
Preserve purged query vars during canonical redirect

This supercedes 79aeeb1

westonruter committed:
Preserve purged query args on all redirects not just canonical ones

westonruter committed:
Use AMP shadow streaming API with fetch intead of XHR

westonruter committed:
Add Web Components polyfill if Shadow DOM is not natively available

miina committed:
Use IIFE to support Safari

westonruter committed:
Remove unused fetchDocument function

westonruter committed:
Remove unused amp_script_attributes data; allow amp_script_attributes for non-AMP CDS assets

westonruter committed:
Add amp-wp-app-shell to precache and use md5 as revision when WP_DEBUG

westonruter committed:
Force AMP endpoint whenever requesting inner app shell

westonruter committed:
Export isOuterAppShell to page

westonruter committed:
Use Loading as document title for app shell

westonruter committed:
Fix serving error templates for app shell

westonruter committed:
Fix adding current nav menu item class on initial app shell load

westonruter committed:
Add service worker code removed from PWA plugin in https://github.com/xwp/pwa-wp/pull/107/commits/bd9084db000d5687d3a75f042246b6b05b6bfe56

westonruter committed:
Make sure offline template is included among supportable templates

westonruter commented on assets/js/amp-service-worker-offline-commenting.js:
```diff
@@ -0,0 +1,45 @@
+					body: body
+				} );
+
+				// Add request to queue. @todo Replace when upgrading to Workbox v4!
```
@miina Is this still valid?

miina replied:
@westonruter I believe so, that was related to https://github.com/GoogleChrome/workbox/pull/1710 specifically. Looks like it got merged meanwhile.

westonruter replied:
It's required sure enough. The offline commenting is currently failing:

![image](https://user-images.githubusercontent.com/134745/49178146-d8a5dc00-f303-11e8-92ef-7c29b4ba87ad.png)

westonruter replied:
Fixed in https://github.com/Automattic/amp-wp/pull/1519/commits/1e77c6e56f3ad0b8a39a988f4aab409bb646e326 and https://github.com/xwp/pwa-wp/pull/109

westonruter replied:
Correction: https://github.com/Automattic/amp-wp/pull/1519/commits/ee465e718f52f6ea90d102a9b033240aa24505a6

westonruter committed:
Update todo comments

westonruter committed:
Fix background sync for offline comments in Workbox 4

westonruter committed:
Fix passing request to queue.pushRequest

westonruter commented:
We'll need to merge the updates to the underlying service worker code in #1261, as it will be merged before this is.

miina committed:
Merge develop & resolve conflicts.

miina committed:
Update ignoring phpcs rule.

miina committed:
Fix more phpcs issues.

indieisaconcept committed:
Ensure .site-content-contain element exists 

**Rationale**
A prior change introduced an early return for `isHeaderVisible` should the
`site-branding` element not be found in the document. However this early
return caused code to trigger elsewhere related to scrolling.

**Changes**

- Ensure `.site-content-contain` document lookup is also guarded to prevent
  calling scrollIntoView on a non-existing dom element.

mehigh committed:
Merge branch 'develop' into add/spa-amp-shadow-dom

# Conflicts:
#	.eslintrc
#	amp.php
#	assets/js/amp-service-worker-runtime-precaching.js
#	includes/amp-helper-functions.php
#	includes/class-amp-http.php
#	includes/class-amp-service-worker.php
#	includes/class-amp-theme-support.php
#	includes/sanitizers/class-amp-style-sanitizer.php

mehigh commented:
@miina 
I've handled the 8 conflicts and issued a PR against this branch here which handles them:
https://github.com/ampproject/amp-wp/pull/3931 
This is an effort towards getting it closer to being merged.

mehigh committed:
AppShell polyfill: Update to latest version - 2.4.0

mehigh committed:
Amp-Wp-App-Shell - Improve flexibility for different types of caching

Update amp-wp-app-shell.js

**Rationale**
Due to caching restrictions with WPengine the use of a query-string parameter to
denote an inner shell request is problematic as it prevents the request from
being cached independently. The request needs to be distinctly different from
the outer shell in-order for it to be treated as a seperate cache item.

**Changes**

- Query string parameter is now appended to the path instead of
- Fixed bug where code was always expecting a .site-branding element to exist

mehigh committed:
Add a filter to allow altering AMP output from themes

westonruter committed:
Add missing changes to style sanitizer; remove duplicated autoloader entry

westonruter committed:
Merge branch 'develop' of github.com:ampproject/amp-wp into add/spa-amp-shadow-dom-updated

* 'develop' of github.com:ampproject/amp-wp: (233 commits)
  Ignore Story editor tests (#4043)
  Update dependency eslint-plugin-jest to v23.4.0 (#4056)
  Update dependency xwp/wp-dev-lib to v1.4.0
  Update amp-video embed regex pattern to include other Vimeo URL formats (#4051)
  Update amp-instagram embed regex (#4053)
  Update dependency terser-webpack-plugin to v2.3.2 (#4047)
  Fix ability to update the status of a validation error
  Update wp-dev-lib package (#4029)
  Update dependency core-js to v3.6.2 (#4028)
  Update dependency @wordpress/element to v2.10.0 (#3986)
  Update dependency browserslist to v4.8.3 (#4021)
  Update dependency grunt-http to v2.3.3 (#4022)
  Update dependency postcss to v7.0.26 (#4017)
  Update dependency eslint-utils to v2 (#4012)
  Update dependency eslint-plugin-jest to v23.3.0 (#4016)
  Update dependency css-loader to v3.4.1 (#4023)
  Update dependency webpack to v4.41.5 (#4014)
  Update dependency core-js to v3.6.1 (#4011)
  Update dependency eslint to v6.8.0 (#4002)
  Swap PHP 5.4 for 5.6 to run external-http tests
  ...

westonruter committed:
Fix phpcs linting issues

westonruter committed:
Restore missing live_list_offline_commenting to AMP_Service_Worker

westonruter committed:
Make use of new Amp\AmpWP\Document methods; add line breaks

westonruter committed:
Fix build and unit tests by preserving uncompiled scripts

westonruter committed:
Fix eslint issues

westonruter committed:
Merge pull request #3931 from xwp/add/spa-amp-shadow-dom-updated

Update add/spa-amp-shadow-dom with latest in develop

mehigh committed:
AppShell polyfill: Update to latest version - 2.4.1

dero commented on assets/js/amp-wp-app-shell.js:
```diff
@@ -0,0 +1,409 @@
+				 * https://www.ampproject.org/latest/blog/streaming-in-the-shadow-reader/
+				 * https://github.com/ampproject/amphtml/blob/master/spec/amp-shadow-doc.md#fetching-and-attaching-shadow-docs
+				 */
+				currentShadowDoc = AMP.attachShadowDocAsStream( newContainer, url, {} );
```
@westonruter The `url` parameter here comes through as an [URL object](https://developer.mozilla.org/en-US/docs/Web/API/URL) when the navigation happens through a click or form submit, but the method [expects](https://github.com/ampproject/amphtml/blob/f41de5e5205ae37a823e774af817e2748ad5ed09/src/multidoc-manager.js#L273) and [newly enforces](https://github.com/ampproject/amphtml/commit/f41de5e5205ae37a823e774af817e2748ad5ed09) the parameter to be a `string`.

This newly leads to an error being raised now, which breaks regular navigation and potentially form submissions.

westonruter replied:
So this should be `url.toString()`?

westonruter replied:
Feel free to open a PR to this branch!

dero replied:
Opened https://github.com/ampproject/amp-wp/pull/4261

dero committed:
attachShadowDocAsStream expects a string type variable

A recent change on the AMP platform enforces the URL parameter
to be a string, but a URL object instance is passed on navigation
or form submission currently. Adding an explicit type cast fixes
the issue.

ref. https://github.com/ampproject/amphtml/blob/f41de5e5205ae37a823e774af817e2748ad5ed09/src/multidoc-manager.js#L273
ref. https://github.com/ampproject/amphtml/commit/f41de5e5205ae37a823e774af817e2748ad5ed09

westonruter committed:
Merge pull request #4261 from dero/add/spa-amp-shadow-dom

attachShadowDocAsStream expects a string type variable

westonruter committed:
Revert introduction of amp_document_output filter for now

westonruter committed:
Merge pull request #4055 from xwp/add/spa-amp-shadow-dom-extras

SPA with amp shadow DOM extras

delawski commented on includes/amp-helper-functions.php:
```diff
@@ -1228,6 +1247,63 @@ function amp_wp_kses_mustache( $markup ) {
+		ob_end_clean();
+	}
+
+	echo '</div><!-- #amp-app-shell-content -->';
```
```suggestion
	printf( '</div><!-- #%s -->', esc_attr( AMP_Theme_Support::APP_SHELL_CONTENT_ELEMENT_ID ) );
```

delawski commented on includes/class-amp-theme-support.php:
```diff
@@ -190,13 +205,16 @@ public static function init() {
+			add_action( 'parse_query', [ __CLASS__, 'init_app_shell' ], 9 );
 
 			/*
 			 * Note that wp action is use instead of template_redirect because some themes/plugins output
 			 * the response at this action and then short-circuit with exit. So this is why the the preceding
```
```suggestion
			 * Note that wp action is used instead of template_redirect because some themes/plugins output
			 * the response at this action and then short-circuit with exit. So this is why the preceding
```

delawski commented on includes/class-amp-theme-support.php:
```diff
@@ -456,6 +474,70 @@ static function() {
+						return;
+					}
+					wp_die(
+						esc_html__( 'Outer app shell can only be requested of the non-AMP version (thus requires paired mode).', 'amp' ),
```
```suggestion
						esc_html__( 'Outer app shell can only be requested of the non-AMP version (thus requires Transitional mode).', 'amp' ),
```

westonruter committed:
Reuse constant

Co-authored-by: Piotr Delawski <delawski@users.noreply.github.com>

westonruter committed:
Improve copy in comment and error message

Co-authored-by: Piotr Delawski <delawski@users.noreply.github.com>

delawski committed:
Compile app shell JS as part of the webpack build process

Building the app shell JS with webpack allows us to use third-party code
(e.g. @wordpress packages) easily. We also take advantage of the code
minification and babel transpilation.

This is a step towards using @wordpress/hooks package in the app shell
script.

delawski committed:
Use query param for inner component URL (revert 6054285f382b60790a0a4ab8af4fcd550959ec41)

The change to the inner component URL format introduced in 6054285f382b60790a0a4ab8af4fcd550959ec41
broke the app shell experience. While URL format (use an URL segment
instead of a query parameter) was changed in the JS implementation, the
back-end PHP handler was still expecting the old format. It resulted in
loading 404 pages instead of actual inner components for requested pages.

The follow up commit will allow altering of the URL format if a plugin
integrator really needs to do so.

delawski committed:
Allow filter of the inner component URL

In some cases, the plugin integrator may need to alter the format of the
inner component URL.

While it is possible to change the way incoming requests for inner
components are handled on the back-end (PHP), there was no way to change
the URL format of the request on the front-end (JS).

This commit makes the inner component request URL filterable using
wp.hooks. Here's how you could filter the URL in your plugin or theme
so that instead of using a query parameter, a new segment is added to
the URL:

```
if ( ampAppShell && ampAppShell.hooks ) {
	function filterInnerComponentUrl( url, baseUrl, componentQueryVar ) {
		const filteredUrl = new URL( baseUrl );
		const pathSuffix = '_' + componentQueryVar + '_inner';
		filteredUrl.pathname = filteredUrl.pathname + pathSuffix;

		return filteredUrl;
	}

	ampAppShell.hooks.addFilter( 'amp.appShell.innerComponentUrl', 'ampWpAppShell/filterInnerComponentUrl', filterInnerComponentUrl );
}
```

delawski committed:
Fix lint issues

westonruter committed:
Merge pull request #5087 from xwp/fix/spa-amp-fetch-inner

Allow filtering of inner component URL

delawski committed:
Merge branch 'develop' into add/spa-amp-shadow-dom

* develop: (2551 commits)
  Update dependency eslint-plugin-jest to v23.18.2
  Update dependency eslint-plugin-jest to v23.18.1
  Fix TikTok test
  Fix test failure in WP 4.9
  Add tests for ObsoleteBlockAttributeRemover
  Skip adding data attributes for gallery shortcodes
  Revert block deprecation logic
  Harden pattern to match block comment and start tag
  Improve phpdoc for ObsoleteBlockAttributeRemover
  Improve logic for obtaining obsolete attribute pattern
  Introduce ObsoleteBlockAttributeRemover service
  Update dependency webpack to v4.44.0
  Modify the block attributes during its deprecation
  Use DOMElement as return type for HasCaption::get_caption_element()
  Prevent removal of data attributes in sanitize_raw_embeds and fix boolean checks
  Ensure data-amp-carousel=true attribute is present for proper styling of gallery block
  Supply missing attribute value for gallery shortcode output
  Ensure comments_popup_link() is also filtered to prevent mobile redirection
  Test AMP_Theme_Support::amend_comments_link
  Remove unused imports
  ...

# Conflicts:
#	amp.php
#	includes/amp-helper-functions.php
#	includes/class-amp-theme-support.php
#	includes/sanitizers/class-amp-style-sanitizer.php
#	includes/sanitizers/class-amp-tag-and-attribute-sanitizer.php

delawski committed:
Fix indentation

delawski committed:
Use proper method name

delawski committed:
Fix lint issues

westonruter committed:
Merge pull request #5117 from xwp/add/spa-amp-shadow-dom

Merge develop and resolve conflicts

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/ampproject/amp-wp?pullRequest=1519) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/ampproject/amp-wp?pullRequest=1519) before we can accept your contribution.<br/>**4** out of **6** committers have signed the CLA.<br/><br/>:white_check_mark: westonruter<br/>:white_check_mark: dero<br/>:white_check_mark: delawski<br/>:white_check_mark: mehigh<br/>:x: miina<br/>:x: indieisaconcept<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/ampproject/amp-wp?pullRequest=1519) it.</sub>
